File: pages/api/insurance.ts
import { NextApiRequest, NextApiResponse } from 'next';
import { predictInsuranceCost, InsuranceData } from './ml-utils';

interface InsuranceInput {
  age: number;
  sex: string;
  bmi: number;
  children: number;
  smoker: string;
  region: string;
}

interface InsuranceResponse {
  predictedCost: number;
  currency: string;
  modelType: string;
  inputData: InsuranceInput;
}

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<InsuranceResponse | { error: string } | { message: string }>
) {
  // Handle CORS
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method === 'GET') {
    res.status(200).json({ message: 'Insurance cost prediction API is running (Native JS ML Model)' });
    return;
  }

  if (req.method === 'POST') {
    try {
      const {
        age = 0,
        sex = 'male',
        bmi = 0,
        children = 0,
        smoker = 'no',
        region = 'northeast'
      } = req.body as InsuranceInput;

      // Prepare data for ML model (only features used in training)
      const modelInput: InsuranceData = {
        age,
        bmi,
        children,
        smoker
      };

      // Make prediction using native JavaScript
      const predictedCost = predictInsuranceCost(modelInput);

      const response: InsuranceResponse = {
        predictedCost: Math.round(predictedCost * 100) / 100, // Round to 2 decimal places
        currency: 'USD',
        modelType: 'Native JS ML Model (Polynomial Regression)',
        inputData: {
          age,
          sex,
          bmi,
          children,
          smoker,
          region
        }
      };

      res.status(200).json(response);

    } catch (error) {
      res.status(500).json({ error: error instanceof Error ? error.message : 'Unknown error' });
    }
  } else {
    res.status(405).json({ error: 'Method not allowed' });
  }
}

File: api/insurance.py
from http.server import BaseHTTPRequestHandler
import json
import os
import sys
import joblib
import pandas as pd

# Add the scripts directory to path
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
scripts_dir = os.path.join(parent_dir, 'scripts')
sys.path.append(scripts_dir)

# Import custom transformers
try:
    from insurance_transformers import CustomLabelEncoder, FeatureSelector
except ImportError:
    # If transformers not available, we'll handle it gracefully
    CustomLabelEncoder = None
    FeatureSelector = None

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()

        response = {
            "message": "Insurance cost prediction API is running (Python ML Model)"
        }
        self.wfile.write(json.dumps(response).encode())

    def do_POST(self):
        try:
            # Read request body
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            data = json.loads(post_data.decode('utf-8'))

            # Extract input parameters with defaults
            age = data.get('age', 0)
            sex = data.get('sex', 'male')
            bmi = data.get('bmi', 0)
            children = data.get('children', 0)
            smoker = data.get('smoker', 'no')
            region = data.get('region', 'northeast')

            # Load the model
            model_path = os.path.join(parent_dir, 'insurance_cost_model.pkl')

            if not os.path.exists(model_path):
                raise FileNotFoundError(f"Model file not found at {model_path}")

            model = joblib.load(model_path)

            # Create DataFrame with the features used in training
            # Based on the model, only age, bmi, children, smoker were used
            input_data = pd.DataFrame({
                'age': [int(age)],
                'bmi': [float(bmi)],
                'children': [int(children)],
                'smoker': [smoker]  # 'yes' or 'no'
            })

            # Make prediction
            predicted_cost = model.predict(input_data)[0]

            # Prepare response
            response = {
                "predictedCost": round(float(predicted_cost), 2),
                "currency": "USD",
                "modelType": "Python ML Model (Polynomial Regression)",
                "inputData": {
                    "age": age,
                    "sex": sex,
                    "bmi": bmi,
                    "children": children,
                    "smoker": smoker,
                    "region": region
                }
            }

            # Send response
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(response).encode())

        except Exception as e:
            # Send error response
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()

            error_response = {
                "error": str(e)
            }
            self.wfile.write(json.dumps(error_response).encode())

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()